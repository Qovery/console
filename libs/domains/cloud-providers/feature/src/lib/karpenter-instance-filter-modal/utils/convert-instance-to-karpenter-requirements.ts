import {
  type ClusterInstanceTypeResponseListResultsInner,
  type KarpenterNodePoolRequirement,
} from 'qovery-typescript-axios'

export function convertToKarpenterRequirements(
  instance: ClusterInstanceTypeResponseListResultsInner[]
): KarpenterNodePoolRequirement[] {
  if (!instance || instance.length === 0) {
    return []
  }

  const requirements: KarpenterNodePoolRequirement[] = []

  const uniqueValues = {
    architectures: new Set<string>(),
    families: new Set<string>(),
    sizes: new Set<string>(),
  }

  instance.forEach((instance) => {
    if (instance.architecture) {
      uniqueValues.architectures.add(instance.architecture)
    }

    if (instance.attributes?.instance_family) {
      uniqueValues.families.add(instance.attributes.instance_family)
    }
    if (instance.attributes?.instance_size) {
      uniqueValues.sizes.add(instance.attributes.instance_size)
    }
  })

  if (uniqueValues.sizes.size > 0) {
    requirements.push({
      key: 'InstanceSize',
      operator: 'In',
      values: Array.from(uniqueValues.sizes),
    })
  }

  if (uniqueValues.architectures.size > 0) {
    requirements.push({
      key: 'Arch',
      operator: 'In',
      values: Array.from(uniqueValues.architectures),
    })
  }

  if (uniqueValues.families.size > 0) {
    requirements.push({
      key: 'InstanceFamily',
      operator: 'In',
      values: Array.from(uniqueValues.families),
    })
  }

  return requirements
}
