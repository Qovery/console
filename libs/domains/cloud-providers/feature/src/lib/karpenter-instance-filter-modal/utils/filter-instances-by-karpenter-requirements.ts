import { type ClusterInstanceTypeResponseListResultsInner } from 'qovery-typescript-axios'
import { type KarpenterData } from '@qovery/shared/interfaces'

export function filterInstancesByKarpenterRequirements(
  allInstances: ClusterInstanceTypeResponseListResultsInner[],
  karpenterData: Omit<KarpenterData, 'disk_size_in_gib' | 'enabled' | 'spot_enabled'>
): ClusterInstanceTypeResponseListResultsInner[] {
  const karpenterRequirements = karpenterData.qovery_node_pools?.requirements || []

  return allInstances.filter((instance) => {
    const instanceSize = instance.attributes?.instance_size || ''
    const InstanceFamily = instance.attributes?.instance_family || ''
    const instanceArchitecture = instance.architecture || ''

    return karpenterRequirements.every((requirement) => {
      if (requirement.key === 'InstanceSize') {
        return requirement.values.includes(instanceSize)
      }

      if (requirement.key === 'Arch') {
        // Fix inconsistency between uppercase and lowercase with architecture during the migration
        return requirement.values.map((v) => v.toUpperCase()).includes(instanceArchitecture)
      }

      if (requirement.key === 'InstanceFamily') {
        return requirement.values.includes(InstanceFamily)
      }

      return false
    })
  })
}

export default filterInstancesByKarpenterRequirements
