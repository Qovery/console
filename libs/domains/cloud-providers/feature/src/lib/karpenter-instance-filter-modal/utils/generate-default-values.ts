import { type ClusterInstanceTypeResponseListResultsInner } from 'qovery-typescript-axios'
import { type KarpenterInstanceFormProps } from '../karpenter-instance-filter-modal'

export function generateDefaultValues(
  instances: ClusterInstanceTypeResponseListResultsInner[]
): Omit<KarpenterInstanceFormProps, 'default_service_architecture'> {
  // Get unique architectures
  const architectures = [...new Set(instances.map((i) => i.architecture))]

  // Get unique sizes without sorting
  const sizes = [...new Set(instances.map((i) => i.attributes?.instance_size ?? ''))].filter((i) => i !== '')

  // Group instance families by raw instance_category
  const categories: Record<string, string[]> = {}
  instances.forEach((instance) => {
    const family = instance.attributes?.instance_family || ''
    const category = instance.attributes?.instance_category || ''

    if (!categories[category]) {
      categories[category] = []
    }

    if (!categories[category].includes(family)) {
      categories[category].push(family)
    }
  })

  return {
    ARM64: architectures.includes('ARM64'),
    AMD64: architectures.includes('AMD64'),
    categories,
    sizes,
  }
}

export default generateDefaultValues
