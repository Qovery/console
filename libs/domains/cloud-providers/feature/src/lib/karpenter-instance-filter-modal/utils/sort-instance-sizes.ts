type BasicSize = 'small' | 'medium' | 'large' | 'xlarge'
type SizePriority = Record<BasicSize, number>

// Define type for the sort key tuple
type SortKey = [priority: number, multiplier: number]

export function sortInstanceSizes(sizes: string[]): string[] {
  const sizePriority: SizePriority = {
    small: 1,
    medium: 2,
    large: 3,
    xlarge: 4,
  }

  function getSortKey(size: string): SortKey {
    if (isBasicSize(size)) {
      return [sizePriority[size], 0]
    }

    // Handle xlarge variants
    if (size.includes('xlarge')) {
      // Extract number before 'xlarge'
      const number = size.replace('xlarge', '')
      // Convert empty string to 1 (for just 'xlarge')
      const multiplier = number ? parseInt(number) : 1
      return [5, multiplier]
    }

    return [0, 0]
  }

  function isBasicSize(size: string): size is BasicSize {
    return size in sizePriority
  }

  return sizes.sort((a, b) => {
    const [priorityA, multiplierA] = getSortKey(a)
    const [priorityB, multiplierB] = getSortKey(b)

    if (priorityA !== priorityB) {
      return priorityA - priorityB
    }
    return multiplierA - multiplierB
  })
}

export default sortInstanceSizes
