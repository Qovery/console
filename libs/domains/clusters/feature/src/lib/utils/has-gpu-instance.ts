import { type Cluster } from 'qovery-typescript-axios'

export const hasGpuInstance = (cluster?: Cluster) => {
  const clusterFeatureKarpenter = cluster?.features?.find((feature) => feature.id === 'KARPENTER')
  if (!clusterFeatureKarpenter) return false
  return Boolean(
    typeof clusterFeatureKarpenter?.value_object?.value === 'object' &&
      'qovery_node_pools' in clusterFeatureKarpenter.value_object.value &&
      clusterFeatureKarpenter.value_object.value.qovery_node_pools.gpu_override
  )
}
