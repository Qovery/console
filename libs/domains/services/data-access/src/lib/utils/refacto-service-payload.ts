import {
  type ApplicationEditRequest,
  type ApplicationGitRepositoryRequest,
  DatabaseModeEnum,
  type HelmRequest,
  type HelmResponseAllOfSourceOneOf,
  type HelmResponseAllOfSourceOneOf1,
  type ServiceStorageStorageInner,
} from 'qovery-typescript-axios'
import { match } from 'ts-pattern'
import { isJobGitSource } from '@qovery/shared/enums'
import { type Application, type Container, type Database, type Helm, type Job } from '../domains-services-data-access'

/* 
TODO: all this following functions should be removed after the API refactoring and we need to clean it after the Services migration to React Query 
https://www.notion.so/qovery/API-improvements-b54ba305c2ee4e549eb002278c532c7f?pvs=4#7d71ea23b5fa44ca80c15a0c32ebd8da
*/

export function refactoApplication(application: Application) {
  // refacto because we can't send all git data
  if (application.git_repository) {
    application.git_repository = {
      url: application.git_repository.url,
      branch: application.git_repository.branch,
      root_path: application.git_repository.root_path,
      git_token_id: application.git_repository.git_token_id,
    }
  }

  // refacto to remove the id by storage
  if (application.storage) {
    application.storage =
      application.storage.length > 0
        ? application.storage.map((storage: ServiceStorageStorageInner) => ({
            mount_point: storage.mount_point,
            size: storage.size,
            type: storage.type,
            id: storage.id,
          }))
        : []
  }

  const applicationRequestPayload: ApplicationEditRequest = {
    name: application.name,
    storage: application.storage,
    cpu: application.cpu,
    git_repository: application.git_repository as ApplicationGitRepositoryRequest,
    build_mode: application.build_mode,
    description: application.description || undefined,
    memory: application.memory,
    auto_preview: application.auto_preview,
    auto_deploy: application.auto_deploy,
    ports: application.ports,
    dockerfile_path: application.dockerfile_path || undefined,
    healthchecks: application.healthchecks ?? {},
    buildpack_language: application.buildpack_language,
    max_running_instances: application.max_running_instances,
    min_running_instances: application.min_running_instances,
    entrypoint: application.entrypoint,
    arguments: application.arguments,
  }

  return applicationRequestPayload
}

export function refactoContainer(container: Container) {
  const containerRequestPayload = {
    name: container.name || '',
    description: container.description || '',
    storage: container.storage,
    ports: container.ports,
    cpu: container.cpu,
    memory: container.memory,
    max_running_instances: container.max_running_instances,
    min_running_instances: container.min_running_instances,
    registry_id: container.registry?.id || '',
    image_name: container.image_name || '',
    tag: container.tag || '',
    arguments: container.arguments,
    entrypoint: container.entrypoint,
    auto_preview: container.auto_preview,
    auto_deploy: container.auto_deploy,
    healthchecks: container.healthchecks,
  }

  return containerRequestPayload
}

export function refactoJob(job: Job) {
  const jobRequest = {
    name: job.name,
    description: job.description,
    cpu: job.cpu,
    memory: job.memory,
    auto_preview: job.auto_preview,
    auto_deploy: job.auto_deploy,
    max_duration_seconds: job.max_duration_seconds,
    port: job.port,
    max_nb_restart: job.max_nb_restart,
    schedule: job.schedule,
    healthchecks: job.healthchecks ?? {},
    source: {},
  }

  if (isJobGitSource(job.source)) {
    jobRequest.source = {
      docker: {
        dockerfile_path: job.source.docker?.dockerfile_path,
        git_repository: {
          url: job.source.docker?.git_repository?.url || '',
          branch: job.source.docker?.git_repository?.branch,
          root_path: job.source.docker?.git_repository?.root_path,
          git_token_id: job.source.docker?.git_repository?.git_token_id,
        },
      },
    }
  } else {
    jobRequest.source = {
      image: {
        registry_id: job.source?.image?.registry_id,
        image_name: job.source?.image?.image_name,
        tag: job.source?.image?.tag,
      },
    }
  }

  return jobRequest
}

export function refactoDatabase(database: Database) {
  const databaseRequestPayload = {
    name: database.name,
    description: database.description,
    version: database.version,
    accessibility: database.accessibility,
    cpu: database.cpu,
    memory: database.memory,
    storage: database.storage,
    instance_type: database.instance_type,
  }

  if (database.mode === DatabaseModeEnum.MANAGED) databaseRequestPayload.version = database.version

  return databaseRequestPayload
}

export function refactoHelm(helm: Helm) {
  const sourceProvider = (helm?.source as HelmResponseAllOfSourceOneOf).git ? 'GIT' : 'HELM_REPOSITORY'

  const source = match(sourceProvider)
    .with('GIT', () => {
      const gitRepository = (helm?.source as HelmResponseAllOfSourceOneOf).git?.git_repository
      return {
        git_repository: {
          url: gitRepository?.url ?? '',
          branch: gitRepository?.branch,
          root_path: gitRepository?.root_path,
          git_token_id: gitRepository?.git_token_id,
        },
      }
    })
    .with('HELM_REPOSITORY', () => {
      const helmRepository = (helm?.source as HelmResponseAllOfSourceOneOf1).repository
      return {
        helm_repository: {
          repository: helmRepository?.repository?.id,
          chart_name: helmRepository?.chart_name,
          chart_version: helmRepository?.chart_version,
        },
      }
    })
    .exhaustive()

  const helmRequestPayload: HelmRequest = {
    name: helm.name,
    description: helm.description,
    auto_preview: helm.auto_preview,
    auto_deploy: helm.auto_deploy,
    ports: helm.ports,
    allow_cluster_wide_resources: helm.allow_cluster_wide_resources,
    arguments: helm.arguments,
    values_override: helm.values_override,
    source,
  }

  return helmRequestPayload
}
