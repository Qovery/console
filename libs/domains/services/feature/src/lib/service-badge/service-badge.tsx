import { type VariantProps, cva } from 'class-variance-authority'
import { type JobLifecycleTypeEnum } from 'qovery-typescript-axios'
import { type ComponentPropsWithoutRef, type ElementRef, forwardRef } from 'react'
import { type IconEnum } from '@qovery/shared/enums'
import { Icon, Tooltip } from '@qovery/shared/ui'
import { twMerge, upperCaseFirstLetter } from '@qovery/shared/util-js'

const serviceBadgeVariants = cva(
  ['relative', 'flex', 'items-center', 'justify-center', 'rounded-full', 'border', 'border-neutral-200'],
  {
    variants: {
      size: {
        md: ['h-16', 'w-16'],
        xs: ['h-8', 'w-8'],
      },
    },
  }
)

const iconVariants = cva('', {
  variants: {
    size: {
      md: ['h-10', 'w-10'],
      xs: ['h-6', 'w-6'],
    },
  },
})

const typeVariants = cva(['absolute', '-right-0.5', '-top-0.5', 'p-0.5'], {
  variants: {
    size: {
      md: ['h-6', 'w-6'],
      xs: ['h-4', 'w-4', '-top-1', '-right-1.5'],
    },
  },
})

export interface ServiceBadgeProps
  extends Omit<ComponentPropsWithoutRef<'div'>, 'size'>,
    VariantProps<typeof serviceBadgeVariants> {
  icon: keyof typeof IconEnum
  type?: keyof typeof JobLifecycleTypeEnum
}

// TODO: Prepare component to update service icon with custom icon
export const ServiceBadge = forwardRef<ElementRef<'div'>, ServiceBadgeProps>(function ServiceBadge(
  { className, size = 'md', icon, type },
  forwardedRef
) {
  return (
    <div className={twMerge(serviceBadgeVariants({ size }), className)} ref={forwardedRef}>
      <Icon className={twMerge(iconVariants({ size }))} name={icon} />
      {type && (
        <Tooltip content={`Deployed with ${upperCaseFirstLetter(type)}`} disabled={size === 'xs'}>
          <span className={twMerge(typeVariants({ size }))}>
            <Icon className="h-full w-full" name={type} />
          </span>
        </Tooltip>
      )}
    </div>
  )
})

export default ServiceBadge
