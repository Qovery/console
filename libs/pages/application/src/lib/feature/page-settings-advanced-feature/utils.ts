import { ApplicationAdvancedSettings } from 'qovery-typescript-axios'
import { ApplicationEntity } from '@console/shared/interfaces'

export function initFormValues(keys: string[], application: ApplicationEntity): { [key: string]: string } {
  const values: { [key: string]: string } = {}
  keys.forEach((key) => {
    if (application.advanced_settings?.current_settings) {
      values[key] =
        application.advanced_settings.current_settings[key as keyof ApplicationAdvancedSettings]?.toString() || ''
    }
  })

  return values
}
