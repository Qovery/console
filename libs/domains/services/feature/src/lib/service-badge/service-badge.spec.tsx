import { wrapWithReactHookForm } from '__tests__/utils/wrap-with-react-hook-form'
import { renderWithProviders } from '@qovery/shared/util-tests'
import { ServiceBadge } from './service-badge'

describe('ServiceBadge', () => {
  it('should match snapshot', () => {
    const { baseElement } = renderWithProviders(
      wrapWithReactHookForm(<ServiceBadge size="md" icon="LIFECYCLE_JOB" type="TERRAFORM" />)
    )
    expect(baseElement).toMatchSnapshot()
  })
})
