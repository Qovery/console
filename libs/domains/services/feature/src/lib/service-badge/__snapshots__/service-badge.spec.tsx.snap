// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceBadge should match snapshot 1`] = `
<body>
  <div>
    <div
      class="relative flex items-center justify-center rounded-full border border-neutral-200 h-16 w-16"
    >
      <svg
        class="shrink-0 h-10 w-10"
        fill="none"
        name="LIFECYCLE_JOB"
        viewBox="0 0 22 22"
        width="1.5rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          opacity="0.7"
        >
          <circle
            cx="10.875"
            cy="11"
            r="10.15"
            stroke="#42E8E0"
            stroke-width="0.3"
          />
          <ellipse
            cx="10.876"
            cy="11.077"
            fill="#EAEAFE"
            rx="7.669"
            ry="7.669"
          />
          <path
            clip-rule="evenodd"
            d="M1.762 11.076a9.114 9.114 0 019.036-9.113v2.72a6.394 6.394 0 00-6.316 6.393h-2.72z"
            fill="url(#paint0_linear_10128_95071)"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M10.688 1.963a9.218 9.218 0 019.218 9.14h-2.751a6.467 6.467 0 00-6.467-6.389V1.963z"
            fill="url(#paint1_linear_10128_95071)"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M19.906 11a9.231 9.231 0 01-9.036 9.112v-2.756A6.476 6.476 0 0017.15 11h2.756z"
            fill="url(#paint2_linear_10128_95071)"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M10.875 20.113a9.114 9.114 0 01-9.113-9.037h2.72a6.394 6.394 0 006.393 6.317v2.72z"
            fill="url(#paint3_linear_10128_95071)"
            fill-rule="evenodd"
          />
          <path
            d="M11.514 3.326L10.15 1.963v2.726l1.364-1.363zM18.543 11.72l1.363-1.364H17.18l1.363 1.364zM10.15 18.75l1.364 1.363v-2.726L10.15 18.75zM3.152 10.357L1.79 11.72h2.726l-1.363-1.363z"
            fill="#fff"
          />
          <rect
            fill="#fff"
            height="1.252"
            rx="0.626"
            width="5.803"
            x="7.975"
            y="8.274"
          />
          <rect
            fill="#fff"
            height="1.252"
            rx="0.626"
            width="5.803"
            x="7.975"
            y="10.45"
          />
          <rect
            fill="#fff"
            height="1.252"
            rx="0.626"
            width="5.803"
            x="7.975"
            y="12.627"
          />
          <rect
            fill="#8987C9"
            height="1.252"
            rx="0.626"
            width="5.803"
            x="7.975"
            y="8.198"
          />
          <rect
            fill="#8987C9"
            height="1.252"
            rx="0.626"
            width="5.803"
            x="7.975"
            y="10.374"
          />
          <rect
            fill="#8987C9"
            height="1.252"
            rx="0.626"
            width="5.803"
            x="7.975"
            y="12.55"
          />
        </g>
        <defs>
          <lineargradient
            gradientUnits="userSpaceOnUse"
            id="paint0_linear_10128_95071"
            x1="11.349"
            x2="7.509"
            y1="17.65"
            y2="6.193"
          >
            <stop
              stop-color="#8987C9"
            />
            <stop
              offset="1"
              stop-color="#BDBFFF"
            />
          </lineargradient>
          <lineargradient
            gradientUnits="userSpaceOnUse"
            id="paint1_linear_10128_95071"
            x1="16.627"
            x2="13.808"
            y1="11.103"
            y2="4.453"
          >
            <stop
              stop-color="#8987C9"
            />
            <stop
              offset="1"
              stop-color="#BDBFFF"
            />
          </lineargradient>
          <lineargradient
            gradientUnits="userSpaceOnUse"
            id="paint2_linear_10128_95071"
            x1="16.527"
            x2="17.628"
            y1="17.334"
            y2="18.595"
          >
            <stop
              stop-color="#8987C9"
            />
            <stop
              offset="1"
              stop-color="#BDBFFF"
            />
          </lineargradient>
          <lineargradient
            gradientUnits="userSpaceOnUse"
            id="paint3_linear_10128_95071"
            x1="6.85"
            x2="1.762"
            y1="18.01"
            y2="18.19"
          >
            <stop
              stop-color="#8987C9"
            />
            <stop
              offset="1"
              stop-color="#BDBFFF"
            />
          </lineargradient>
        </defs>
      </svg>
      <span
        class="absolute -right-0.5 -top-0.5 p-0.5 h-6 w-6"
        data-state="closed"
      >
        <svg
          class="shrink-0 h-full w-full"
          fill="none"
          name="TERRAFORM"
          viewBox="0 0 11 12"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.28 4.01v3.786l3.28-1.892v-3.79L7.28 4.01zM3.64 2.114L6.92 4.01v3.786L3.64 5.902V2.114zM0 0v3.788l3.28 1.894V1.894L0 0zm3.64 10.106L6.92 12V8.212L3.64 6.318v3.788z"
            fill="#7B42BC"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;
